mewarpx/mewarpx/utils_store/plasma_density_oracle.py

```diff
@@ -0,0 +1,221 @@
+
+        self.diags_dir = os.path.join(work_dir, "diags", "fields")
+        self.output_dir = save_dir
+
+        self.species_list = []
+        for name in species_names:
+            self.species_list.append(
+                {"name": name}
+            )
+
+    def main(self):
+        """Calls all functions in order to get raw data, generate prediction,
+        and save prediction"""
+        self.get_coarse_data()
+        self.generate_regression()
+        self.save_predictions()
+
+    def get_coarse_data(self):
+        """Gets data from diag files and coarsens them"""
+        for species in self.species_list:
+            for i in range(len(self.time_steps)):
+                # pattern match files to allow combining densities from multiple files
+                files = glob.glob(os.path.join(
+                    self.diags_dir,
+                    f"*{species['name']}*_particle_density_{self.time_steps[i]:010d}.npy"
+                ))
+                temp_data = np.load(files[0])
+                for j in range(1, len(files)):
+                    temp_data = temp_data + np.load(files[j])
+
+                # slice off first value to remove ghost cell
+                species[f"data_{i}"] = block_reduce(
+                    temp_data[1:, 1:], self.coarsening_factor, np.mean
+                )
+
+    def generate_regression(self):
+        """Analyzes regressions for each species and calculates the predicted
+        density for each species at the time_skip timestep."""
+        for species in self.species_list:
+            num_rows = species["data_0"].shape[0]
+            num_columns = species["data_0"].shape[1]
+            density_max = np.amax([species["data_0"], species["data_1"], species["data_2"]])
+
+            predict_array = np.zeros_like(species["data_0"])
+
+            for i in range(num_rows):
+                for j in range(num_columns):
+                    # scale data to avoid overflow error
+                    x_data = np.array(self.time_steps) / self.time_skip
+                    y_data = np.array([species["data_0"][i, j],
+                                       species["data_1"][i, j],
+                                       species["data_2"][i, j]]) / density_max
+
+                    # check if data is monotonically increasing or decreasing
+                    if (np.all(y_data[1:] > y_data[:-1]) or
+                       np.all(y_data[1:] < y_data[:-1])):
+                        # use linear regression if plasma density is accelerating
+                        # or if plasma density is linear
+                        if np.abs(y_data[2] - y_data[1]) >= np.abs(y_data[1] - y_data[0]):
+                            m, b = self._solve_linear(x_data, y_data)
+                            # time_skip is at x = 1, due to scaling down
+                            prediction = m + b
+                        # if data is linear or density increase is accelerating
```

Comment:
This is a typo, right?
```suggestion
                        # if data is linear or density increase/decrease is decelerating
```

Reply:
Yep